activation/challenge_verifier.go

```diff
@@ -0,0 +1,155 @@
+	"errors"
+	"fmt"
+
+	atypes "github.com/spacemeshos/go-spacemesh/activation/types"
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/signing"
+)
+
+var (
+	ErrSignatureInvalid = errors.New("signature is invalid")
+	ErrChallengeInvalid = errors.New("challenge is invalid")
+)
+
+type CouldntVerifyError struct {
+	// the source (if any) that caused the error
+	source error
+}
+
+func (e *CouldntVerifyError) Error() string {
+	return fmt.Sprintf("couldn't verify challenge (%v)", e.source)
+}
+
+func (e *CouldntVerifyError) Unwrap() error { return e.source }
+
+func (e *CouldntVerifyError) Is(target error) bool {
+	_, ok := target.(*CouldntVerifyError)
+	return ok
+}
+
+//go:generate mockgen -package=mocks -destination=./mocks/challenge_verifier.go . AtxProvider
+
+type ChallengeVerificationResult struct {
+	Hash   types.Hash32
+	NodeID types.NodeID
+}
+
+type ChallengeVerifier interface {
+	Verify(ctx context.Context, challenge, signature []byte) (*ChallengeVerificationResult, error)
+}
+
+type challengeVerifier struct {
+	atxDB             AtxProvider
+	signatureVerifier signing.VerifyExtractor
+	cfg               atypes.PostConfig
+	goldenATXID       types.ATXID
+	layersPerEpoch    uint32
+}
+
+func NewChallengeVerifier(cdb AtxProvider, signatureVerifier signing.VerifyExtractor, cfg atypes.PostConfig, goldenATX types.ATXID, layersPerEpoch uint32) ChallengeVerifier {
+	return &challengeVerifier{
+		atxDB:             cdb,
+		signatureVerifier: signatureVerifier,
+		cfg:               cfg,
+		goldenATXID:       goldenATX,
+		layersPerEpoch:    layersPerEpoch,
+	}
+}
+
+func (v *challengeVerifier) Verify(ctx context.Context, challengeBytes, signature []byte) (*ChallengeVerificationResult, error) {
+	pubkey, err := v.signatureVerifier.Extract(challengeBytes, signature)
+	if err != nil {
+		return nil, ErrSignatureInvalid
+	}
+	nodeID := types.BytesToNodeID(pubkey.Bytes())
+
+	challenge := types.PoetChallenge{}
+	if err := codec.Decode(challengeBytes, &challenge); err != nil {
+		return nil, err
+	}
+
+	if err := v.verifyChallenge(ctx, &challenge, nodeID); err != nil {
+		return nil, err
+	}
+
+	hash, err := challenge.Hash()
+	if err != nil {
+		return nil, err
+	}
+	return &ChallengeVerificationResult{
+		Hash:   *hash,
+		NodeID: nodeID,
+	}, nil
+}
+
+func (v *challengeVerifier) verifyChallenge(ctx context.Context, challenge *types.PoetChallenge, nodeID types.NodeID) error {
+	log.With().Info("verifying challenge", log.Object("challenge", challenge))
+
+	if err := validateNumUnits(&v.cfg, challenge.NumUnits); err != nil {
+		return fmt.Errorf("%w: %v", ErrChallengeInvalid, err)
+	}
+
+	if err := validatePositioningAtx(&challenge.PositioningATX, v.atxDB, v.goldenATXID, challenge.PubLayerID, v.layersPerEpoch); err != nil {
+		switch err.(type) {
+		case *AtxNotFoundError:
+			return &CouldntVerifyError{source: err}
```

Comment:
i don't understand why use a special error in the case ATX is not found. how does poet server react to two different errors?

Reply:
I want to differentiate between:
- could not verify the challenge,
- the challenge is invalid.

In the former case, Poet will retry to verify the challenge with a different gateway (or the same one, after some time).

The ChallengeVerifier GRPC service logic doesn't know the verification logic and thus is not able to distinguish the errors. For its sake, I wrap every _retryable_ error in `CouldntVerifyError` so that it knows how to respond to the Poet.

Reply:
ideally you can signal retryable error only when the node is not ATX synced. a node is expected to have all the ATXs in the network when it is ATX synced (barring the case of network split).

but i can compromise if you limit the retry attempts in poet server. i just want to avoid this become a dos vector.

Reply:
Poet retries are limited. 

As discussed offline - let's consider distinguishing between synced/not synced and returning invalid/retry status respectively an optimization for the future.